subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java

```diff
@@ -243,6 +261,28 @@ ClasspathFingerprinter createClasspathFingerprinter(ResourceSnapshotterCacheServ
         ClasspathHasher createClasspathHasher(ClasspathFingerprinter fingerprinter, FileCollectionFactory fileCollectionFactory) {
             return new DefaultClasspathHasher(fingerprinter, fileCollectionFactory);
         }
+
+        private static class IncreaseLimitsDocumentation implements InsufficientResourcesForWatchingDocumentationIndex.IncreaseLimitsDocumentation {
+            private final String problem;
+            private final String section;
+            private final DocumentationRegistry documentationRegistry;
+
+            public IncreaseLimitsDocumentation(String problem, String section, DocumentationRegistry documentationRegistry) {
+                this.problem = problem;
+                this.section = section;
+                this.documentationRegistry = documentationRegistry;
+            }
+
+            @Override
+            public String getProblem() {
+                return problem;
+            }
+
+            @Override
+            public String getDocumentationLink() {
+                return documentationRegistry.getDocumentationFor("gradle_daemon", section);
+            }
+        }
     }
```

Comment:
It feels unnecessarily complicated to add so many new classes just to resolve the exceptions. Do you think there's a way to reduce this complexity?

Reply:
I wanted to avoid a dependency on `baseServices` (where `DocumentationRegistry` lives) from `fileWatcher`. I can add the dependency and inject the `DocumentationRegistry` into `WatchingVirtualFileSystem` instead. What is your take on that?

Reply:
I'd like to avoid adding the dependency, too. I wonder if we need both `InsufficientResourcesForWatchingDocumentationIndex` and `IncreaseLimitsDocumentation` as well.

Reply:
I changed things around so we only have one interface. Please take a look.